src/coingecko/Coingecko.ts

```diff
@@ -82,30 +145,74 @@ export class Coingecko {
     currency = "usd",
     platform_id = "ethereum"
   ): Promise<CoinGeckoPrice[]> {
-    // Generate a unique set with no repeated. join the set with the required coingecko delimiter.
-    const contract_addresses = Array.from(new Set(addresses.filter((n) => n).values()));
+    const priceCache: { [addr: string]: CoinGeckoPrice } = this.getPriceCache(currency, platform_id);
+
+    // Pre-populate price cache with requested token addresses
+    addresses.forEach((addr: string) => {
+      if (priceCache[addr] === undefined) {
+        priceCache[addr] = { address: addr, price: 0, timestamp: 0 };
+      }
+    });
```

Comment:
I don't think this causes any obvious issues, but I would suggest doing all the writing synchronously, i.e. don't write before the async CG request and then again after. I would suggest doing all the writing after, so you don't leave the cache in a strange state for others to read while you're waiting on the CG price to come back (if possible).

Reply:
My motivation for pre-populating with any new entries is that it allows the use of priceCache keys as the source for addresses to be requested from CG, so we inherit an array of unique addresses (assuming that the user doesn't request the same token in multiple different cases).  This makes it trivial to generate the list of addresses to be submitted to CoinGecko (i.e. line #158 slightly below this comment).

The existing implementation handled this by creating a temporary set to filter out any duplicate addresses. That's definitely also still valid and achievable; is it preferable to revert to that?

Reply:
No I think this is fine! I generally get a little scared when I see code that crosses awaits, but I do think this is handled okay elsewhere.